### PR TITLE
Fixed issue #17075: User roles : all rights are enabled

### PR DESCRIPTION
**Pull request by Shnoulle**
Dev: clarify functionname in Permission
Dev: use directly DB checking

### CODE REVIEW

**Comment by Shnoulle** [comment]
- hasGlobalPermission :  Returns true if a user has global permission for a certain action.
- hasSurveyPermission : Checks if a user has a certain permission in the given survey

I think hasRolePermission must return true if user have a certain permission in the given role … but not (before this fix)

File: application/models/Permission.php [hunk 1]
```diff
@@ -575,7 +575,7 @@ public function hasPermission($iEntityID, $sEntityName, $sPermission, $sCRUD = '
 
         /* Always return true if you are the owner : this can be done in core plugin ? */
         // TODO: give the rights to owner adding line in permissions table, so it will return true with the normal way
-        if ($iUserID == $this->getEntityOwnerId($iEntityID, $sEntityName)) {
+        if ($iUserID && $iUserID == $this->getEntityOwnerId($iEntityID, $sEntityName)) {
```
Comment (olleharstedt) [inline application/models/Permission.php:578]: Can you be more clear what you test with user id? empty? null? the string "0"?

File: application/models/Permission.php [hunk 2]
```diff
@@ -677,16 +677,26 @@ public function hasSurveyPermission($iSurveyID, $sPermission, $sCRUD = 'read', $
...
-     * @param integer $iUserID integer User ID - if not given the one of the current user is used
-     * @return bool True if user has the permission
+     * Returns true if a role has permission for crud
+     * @param integer $roleId
+     * @param string $sPermission 
+     * @param string $sCRUD The permission detailsyou want to check on: 'create','read','update','delete','import' or 'export'
+     * @return bool allwoedpermssion
      */
-    public function hasRolePermission($iRoleId, $sPermission, $sCRUD = 'read')
+    public function RoleHasPermission($iRoleId, $sPermission, $sCRUD = 'read')
```
Comment (olleharstedt) [inline application/models/Permission.php:686]: Not PSR-12 compliant name, should start with lower letter.

**Review by olleharstedt — CHANGES_REQUESTED** [review]

File: application/models/Permissiontemplates.php [hunk 3]
```diff
@@ -291,9 +291,14 @@ public function removeCdataFormat($node)
         return trim(preg_replace($regex_patterns, $regex_replace, $nodeText));
     }
 
+    /**
+     * return if this role GIVE a permission
+     * @param string $sPermission 
+     * @param string $sCRUD 
+     */
     public function getHasPermission($sPermission, $sCRUD)
```
Comment (Shnoulle) [inline application/models/Permissiontemplates.php:299]: did this function must return if current user have permission ?

Comment (olleharstedt) [reply application/models/Permissiontemplates.php:299]: It sounds like it should return boolean value...?

File: application/models/Permission.php (hunk 2 above)
Comment (Shnoulle) [reply application/models/Permission.php:686]: Oups …

File: application/models/Permissiontemplates.php (hunk 3 above)
Comment (Shnoulle) [reply application/models/Permissiontemplates.php:300]: yes … :) i update phpDoc

Comment (olleharstedt) [reply application/models/Permissiontemplates.php:300]: :+1:

File: application/models/Permission.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/Permission.php]: https://github.com/LimeSurvey/LimeSurvey/pull/1755/commits/5838c394430bcc6093f21eae98eb204c9aad5c1d#diff-10077a3f8791500882779e35a827caf5e6578d5140625094d224de5e3c166780R527

File: application/models/Permission.php [hunk 4]
```diff
@@ -566,16 +566,15 @@ public function hasPermission($iEntityID, $sEntityName, $sPermission, $sCRUD = '
         }
         $sCRUD = $sCRUD . '_p';
 
-        /* Always return false for guests */
-        // TODO: should not be necessary
+        /* Be sure to have an user id */
         $iUserID = $this->getUserId($iUserID);
-        if (!$iUserID && $iUserID !== 0) {
+        /* Always return false for guests */
+        if (!$iUserID) {
```
Comment (olleharstedt) [inline application/models/Permission.php:572]: User id === 0 or === null?

**Review by olleharstedt — CHANGES_REQUESTED** [review]

File: application/models/Permission.php (hunk 4 above)
Comment (Shnoulle) [reply application/models/Permission.php:572]: empty … === 0 or is null or === "" empmty mean empty :)

See https://github.com/LimeSurvey/LimeSurvey/pull/1755/files#diff-10077a3f8791500882779e35a827caf5e6578d5140625094d224de5e3c166780R732

In 3.X : we **never** send 0 to the function …

Comment (olleharstedt) [reply application/models/Permission.php:572]: Writing "empty($userId)" is more clear than "!$userId", I think. Or?

File: application/models/Permissiontemplates.php [hunk 5]
```diff
@@ -291,9 +291,16 @@ public function removeCdataFormat($node)
         return trim(preg_replace($regex_patterns, $regex_replace, $nodeText));
     }
 
+    /**
+     * return if this role GIVE a permission
```
Comment (olleharstedt) [inline application/models/Permissiontemplates.php:295]: Return true?

File: application/models/Permissiontemplates.php [hunk 6]
```diff
@@ -291,9 +291,16 @@ public function removeCdataFormat($node)
...
+     * Used in only in compileExportXML
```
Comment (olleharstedt) [inline application/models/Permissiontemplates.php:296]: One "in" too much.

File: application/models/Permission.php (hunk 4 above)
Comment (Shnoulle) [reply application/models/Permission.php:572]: K

File: application/models/Permissiontemplates.php (hunk 5 above)
Comment (Shnoulle) [reply application/models/Permissiontemplates.php:295]: `* @return boolean` 2 lines after …

File: application/models/Permission.php [hunk 7]
```diff
@@ -611,7 +610,7 @@ public function hasPermission($iEntityID, $sEntityName, $sPermission, $sCRUD = '
             if (safecount($aRoles) > 0) {
                 $allowed = false;
                 foreach ($aRoles as $role) {
-                    $allowed = $allowed || $this->hasRolePermission($role['ptid'], $sPermission, substr($sCRUD, 0, -2));
+                    $allowed = $allowed || $this->RoleHasPermission($role['ptid'], $sPermission, substr($sCRUD, 0, -2));
```
Comment (olleharstedt) [inline application/models/Permission.php:613]: roleHasPermission - R -> r

Comment (Shnoulle) [reply application/models/Permission.php]:
```
$ grep -ri roleHasPermission application/*
models/Permission.php:                    $allowed = $allowed || $this->roleHasPermission($role['ptid'], $sPermission, substr($sCRUD, 0, -2));
models/Permission.php:    public function roleHasPermission($iRoleId, $sPermission, $sCRUD = 'read')
models/Permissiontemplates.php:        return Permission::model()->roleHasPermission($this->ptid, $sPermission, $sCRUD);
views/admin/permissiontemplates/partials/_permissions.php:                                            <?php if(Permission::model()->roleHasPermission($oModel->ptid, $sPermissionKey, $sCRUDKey)):?>

```

**Comment by olleharstedt** [comment]
Thanks, Denis!
